Support basic Opera docker node

Drimix20 opened the pull request:
I would like to use opera node in selenium docker hub. I think that it can be useful also for others, so I created PR to be able to make these changes official and not only for private usage.

I don't have any experience with automation which is created in SeleniumHQ so I tried to make changes in all required other files too, not only created NodeOpera directory.

I tested creation of opera selenium image and it was correctly connected to running selenium docker hub.
I'll provide screenshots and some details about scripts.


- [X] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/docker-selenium/blob/master/CONTRIBUTING.md#contributing-code-to-selenium)

manuferna1985 commented:
I have been configuring a selenium hub using Chrome and Firefox nodes recently, and all this official docker images are simply great! Good work guys! But like Drimix20, I think that if we (me the first) contribute with other images for Opera, Safari, and the other most used browsers that are available for Unix, the % of browsers usage that we could cover would be nearly 90% (ex Edge/IE...).

There is any technical impediment right now to support this project with a Safari nose for example?

Drimix20 commented:
> 
> 
> @Drimix20 Could you please fix the merge conflicts first? After that we can review the PR.

I performed rebase.

Drimix20 commented:
> @manuferna1985
> 
> I have been configuring a selenium hub using Chrome and Firefox nodes recently, and all this official docker images are simply great! Good work guys! But like Drimix20, I think that if we (me the first) contribute with other images for Opera, Safari, and the other most used browsers that are available for Unix, the % of browsers usage that we could cover would be nearly 90% (ex Edge/IE...).
 
> There is any technical impediment right now to support this project with a Safari node for example?
As far as I know there is limitation based on host os system. Currently I don't know how to create docker image with OSX. I saw some hints like "create docker container, install emulator and emulate OSX". But I think this is not good way how to create docker image. Docker image should be lightweight.

diemol commented on Makefile:
```diff
@@ -108,6 +116,8 @@ release_latest:
 	docker push $(NAME)/node-chrome:latest
 	docker push $(NAME)/node-firefox:latest
+	docker push $(NAME)/node-opera:latest
+	docker push $(NAME)/node-phantomjs:latest
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -121,6 +131,8 @@ tag_major_minor:
 	docker tag $(NAME)/node-chrome:$(VERSION) $(NAME)/node-chrome:$(MAJOR)
 	docker tag $(NAME)/node-firefox:$(VERSION) $(NAME)/node-firefox:$(MAJOR)
+	docker tag $(NAME)/node-opera:$(VERSION) $(NAME)/node-opera:$(MAJOR)
+	docker tag $(NAME)/node-phantomjs:$(VERSION) $(NAME)/node-phantomjs:$(MAJOR)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted

diemol commented on Makefile:
```diff
@@ -132,6 +144,8 @@ tag_major_minor:
 	docker tag $(NAME)/node-chrome:$(VERSION) $(NAME)/node-chrome:$(MAJOR).$(MINOR)
 	docker tag $(NAME)/node-firefox:$(VERSION) $(NAME)/node-firefox:$(MAJOR).$(MINOR)
+	docker tag $(NAME)/node-opera:$(VERSION) $(NAME)/node-opera:$(MAJOR).$(MINOR)
+	docker tag $(NAME)/node-phantomjs:$(VERSION) $(NAME)/node-phantomjs:$(MAJOR).$(MINOR)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted

diemol commented on Makefile:
```diff
@@ -143,6 +157,8 @@ tag_major_minor:
 	docker tag $(NAME)/node-chrome:$(VERSION) $(NAME)/node-chrome:$(MAJOR_MINOR_PATCH)
 	docker tag $(NAME)/node-firefox:$(VERSION) $(NAME)/node-firefox:$(MAJOR_MINOR_PATCH)
+	docker tag $(NAME)/node-opera:$(VERSION) $(NAME)/node-opera:$(MAJOR_MINOR_PATCH)
+	docker tag $(NAME)/node-phantomjs:$(VERSION) $(NAME)/node-phantomjs:$(MAJOR_MINOR_PATCH)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted

diemol commented on Makefile:
```diff
@@ -156,6 +172,8 @@ release: tag_major_minor
 	@if ! docker images $(NAME)/node-chrome | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-chrome version $(VERSION) is not yet built. Please run 'make build'"; false; fi
 	@if ! docker images $(NAME)/node-firefox | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-firefox version $(VERSION) is not yet built. Please run 'make build'"; false; fi
+	@if ! docker images $(NAME)/node-opera | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-opera version $(VERSION) is not yet built. Please run 'make build'"; false; fi
+	@if ! docker images $(NAME)/node-phantomjs | awk '{ print $$2 }' | grep -q -F $(VERSION); then echo "$(NAME)/node-phantomjs version $(VERSION) is not yet built. Please run 'make build'"; false; fi
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -167,6 +185,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-chrome:$(VERSION)
 	docker push $(NAME)/node-firefox:$(VERSION)
+	docker push $(NAME)/node-opera:$(VERSION)
+	docker push $(NAME)/node-phantomjs:$(VERSION)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -178,6 +198,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-chrome:$(MAJOR)
 	docker push $(NAME)/node-firefox:$(MAJOR)
+	docker push $(NAME)/node-opera:$(MAJOR)
+	docker push $(NAME)/node-phantomjs:$(MAJOR)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -189,6 +211,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-chrome:$(MAJOR).$(MINOR)
 	docker push $(NAME)/node-firefox:$(MAJOR).$(MINOR)
+	docker push $(NAME)/node-opera:$(MAJOR).$(MINOR)
+	docker push $(NAME)/node-phantomjs:$(MAJOR).$(MINOR)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -200,6 +224,8 @@ release: tag_major_minor
 	docker push $(NAME)/node-chrome:$(MAJOR_MINOR_PATCH)
 	docker push $(NAME)/node-firefox:$(MAJOR_MINOR_PATCH)
+	docker push $(NAME)/node-opera:$(MAJOR_MINOR_PATCH)
+	docker push $(NAME)/node-phantomjs:$(MAJOR_MINOR_PATCH)
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -251,11 +281,15 @@ test_firefox_standalone_debug:
 	firefox \
 	firefox_debug \
+	opera \
+	phantomjs \
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on Makefile:
```diff
@@ -251,11 +281,15 @@ test_firefox_standalone_debug:
 	generate_chrome \
 	generate_firefox \
+	generate_opera \
+	generate_phantomjs \
```
Why is `node-phantomjs` here? This image was deprecated a while ago.

Drimix20 replied:
Deleted.

diemol commented on NodeOpera/Dockerfile.txt:
```diff
@@ -0,0 +1,42 @@
+#=========
+# Opera
+#=========
+ENV OPERA_VERSION=60.0.3255.170
```
Why do we need to hardcode the Opera version, can't the latest one be retrieved dynamically? Can we install any given version?

Drimix20 replied:
Env variable opera version deleted. Always latest available opera will be installed.

diemol commented on NodeOpera/Dockerfile.txt:
```diff
@@ -0,0 +1,42 @@
+# Opera
+#=========
+ENV OPERA_VERSION=60.0.3255.170
+ENV OPERA_DRIVER_VERSION=v.2.45
```
Do we need to hardcode the driver version?

Drimix20 replied:
Deleted. I will use link to latest release instead, so hardcoded version is not required.

diemol commented on NodeOpera/generate_config:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+OPERA_VERSION=60.0.3255.170
```
Why do we need to hardcode the Opera version?

Drimix20 replied:
This hardcoded version is supported by used opera chromium driver. Should I use different approach?

diemol replied:
Please check if the same approach used for Chrome works here, which is installing the latest available package from the distribution. 
The idea is that a browser upgrade should be just running a new build.

Drimix20 replied:
Ok, I removed all hardcoded versions and I made it dynamic based on approach used for Chrome docker images.

diemol commented on Test/smoke-opera.js:
```diff
@@ -0,0 +1 @@
+require('./smoke-test')('opera');
```
There are the old test files when JavaScript was used, please let's use only Python.

Drimix20 replied:
Ok, deleted.

diemol commented on sa-test.sh:
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env bash
```
There are the old test files when JavaScript was used, please let's use only Python.

Drimix20 replied:
Ok, deleted.

diemol commented on test.sh:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env bash
```
There are the old test files when JavaScript was used, please let's use only Python.

Drimix20 replied:
Ok, deleted.

diemol commented on tests/test.py:
```diff
@@ -36,6 +36,9 @@
     'NodeFirefoxDebug': 'node-firefox-debug',
     'StandaloneFirefox': 'standalone-firefox',
     'StandaloneFirefoxDebug': 'standalone-firefox-debug',
+
```
Can you please add tests for the Standalone mode as well?

Drimix20 replied:
Added.

cgoldberg commented on NodeOpera/Dockerfile.txt:
```diff
@@ -0,0 +1,42 @@
+
+RUN apt-get update
+
+RUN apt-get update && \
```
don't need to run `apt-get update` again.

Drimix20 replied:
Deleted.

cgoldberg commented on NodeOpera/Dockerfile.txt:
```diff
@@ -0,0 +1,42 @@
+RUN apt-get update && \
+     apt-get -y --no-install-recommends install opera-stable && \
+     opera --version && \
+     rm -Rf /tmp/* && \
```
why do you need to clean /tmp?

Drimix20 replied:
Deleted. It's leftover from different approach how to install opera.

Drimix20 commented:
@diemol So should I also create standalone and debug version? I created only node version without debug. Thanks for review, I'll resolve all comments.

diemol commented:
@Drimix20 yes, ideally Opera should have images for debug and standalone as well.

Drimix20 commented:
@diemol Yes, I'm already working on others types of docker images. I'm also resolving all comments. Thanks for feedback!

Drimix20 commented:
I added all opera images, I performed rebase and updated tests for other type of images. But right know I'm waiting for PR build, so when it will finish, I'll continue with solving problems with failing tests.

ddavison commented:
related: https://github.com/SeleniumHQ/docker-selenium/pull/791#issuecomment-428666838

I believe we decided against Opera as it uses Chromium as its engine. `NodeChrome` already uses Chromium

Drimix20 commented:
@ddavison  Sorry, I missed your PR, but I was looking for opera keyword.
@diemol Can you please reply?

diemol commented:
@Drimix20 I assumed that you double checked for previous PRs/issues 😄 

I've seen this request for Opera a few times, so instead of saying no this time, I think we should take the approach of accepting the PR, adding Opera, measuring the usage. 
If the images are actually used and helpful, we could continue to support them, but if after a few months we don't see any high usage, we could just remove them.

So far, this PR is not in a good shape, so I would leave it to your judgement @Drimix20 if it makes sense to continue based on the fact that Opera is based on Chromium.

Drimix20 commented:
Ok, I will finish the PR based on approach "accepting the PR, adding Opera, measuring the usage". 

PR is not in good shape because I'm facing to problem with opening new opera session. I found out that there is reported issue https://github.com/operasoftware/operachromiumdriver/issues/69 on operachromiumdriver, that opera version 62.* is not supported by latest operachromiumdriver (2.25). I'm trying to install older 60.* opera version and then I'll see if there is another issue.

I created all type of images so this should be the last missing and important thing.

diemol commented:
Thanks for referecing the issue @Drimix20, which kind of worries me because I would not like to add a browser image if the dependencies are not properly maintained. In this case we cannot use Opera 62 because the driver is not up to date. Maybe this is an exception and the driver is usually well maintained, but this looks like a blocker.

Drimix20 commented:
Hi, operachromedriver is released and is working with Opera 62 version. I updated dockerfile, but I'm still getting error "No matching capabilities found". Can you please navigate me, what can be wrong? 

I've created selenium hub with chrome and opera node. Both nodes were correctly connected to selenium hub. I see also configured capabilities, but I'm not able to find out, why it is not working properly.

Selenium hub with nodes:
![image](https://user-images.githubusercontent.com/7419456/62315061-d1d3f800-b494-11e9-8044-525e58d8e88d.png)


Trying to run script:
`
from selenium import webdriver
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities

driver = webdriver.Remote(
        command_executor='http://127.0.0.1:4444/wd/hub',
        desired_capabilities=DesiredCapabilities.OPERA)
driver.get('http://the-internet.herokuapp.com/nested_frames')
`
